cpp/cmake_modules/DefineOptions.cmake

```diff
@@ -362,7 +362,9 @@ if("${CMAKE_SOURCE_DIR}" STREQUAL "${CMAKE_CURRENT_SOURCE_DIR}")
   define_option(ARROW_WITH_ZSTD "Build with zstd compression" OFF)
 
   define_option(ARROW_WITH_UTF8PROC
-                "Build with support for Unicode properties using the utf8proc library" ON)
+                "Build with support for Unicode properties using the utf8proc library (only used if ARROW_COMPUTE is ON)" ON)
+  define_option(ARROW_WITH_RE2
+    "Build with support for regular expressions using the re2 library (only used if ARROW_COMPUTE or ARROW_GANDIVA is ON)" ON)
```

Comment:
We can remove `ARROW_WITH_UTF8PROC` and `ARROW_WITH_RE2` if we set them in `ThirdpartyToolchain.cmake` internally like `ARROW_WITH_THRIFT`.

Reply:
No, the difference to `ARROW_WITH_THRIFT ` is that you can still build `ARROW_COMPUTE` even if `ARROW_WITH_UTF8PROC=OFF`. In that case, only some features of `compute` won't be build but the majority will work.

Reply:
I understand.